runtime/examples/Rust/libtapasco_tests/src/main.rs

```diff
@@ -0,0 +1,687 @@
+use uom::si::time::microsecond;
+use uom::si::time::nanosecond;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Allocator Error: {}", source))]
+    AllocatorError { source: tapasco::allocator::Error },
+
+    #[snafu(display("DMA Error: {}", source))]
+    DMAError { source: tapasco::dma::Error },
+
+    #[snafu(display("Invalid subcommand"))]
+    UnknownCommand {},
+    #[snafu(display("Failed to initialize TLKM object: {}", source))]
+    TLKMInit { source: tapasco::tlkm::Error },
+
+    #[snafu(display("Failed to decode TLKM device: {}", source))]
+    DeviceInit { source: tapasco::device::Error },
+
+    #[snafu(display("Error while executing Job: {}", source))]
+    JobError { source: tapasco::job::Error },
+
+    #[snafu(display("IO Error: {}", source))]
+    IOError { source: std::io::Error },
+
+    #[snafu(display("Mutex has been poisoned"))]
+    MutexError {},
+
+    #[snafu(display("Crossbeam encountered an error {}", s))]
+    CrossbeamError { s: String },
+}
+
+impl<T> From<std::sync::PoisonError<T>> for Error {
+    fn from(_error: std::sync::PoisonError<T>) -> Self {
+        Error::MutexError {}
+    }
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+fn print_version(_: &ArgMatches) -> Result<()> {
+    let tlkm = TLKM::new().context(TLKMInit {})?;
+    let ver = tlkm.version().context(TLKMInit {})?;
+    println!("TLKM version is {}", ver);
+    Ok(())
+}
+
+fn enum_devices(_: &ArgMatches) -> Result<()> {
+    let tlkm = TLKM::new().context(TLKMInit {})?;
+    let devices = tlkm.device_enum(&HashMap::new()).context(TLKMInit {})?;
+    println!("Got {} devices.", devices.len());
+    for x in devices {
+        println!(
+            "Device {}: Name: {}, Vendor: {}, Product {}",
+            x.id(),
+            x.name(),
+            x.vendor(),
+            x.product()
+        );
+    }
+    Ok(())
+}
+
+fn allocate_devices(_: &ArgMatches) -> Result<()> {
+    let tlkm = TLKM::new().context(TLKMInit {})?;
+    let mut devices = tlkm.device_enum(&HashMap::new()).context(TLKMInit {})?;
+
+    for x in devices.iter_mut() {
+        println!("Allocating ID {} exclusively.", x.id());
+        x.change_access(tapasco::tlkm::tlkm_access::TlkmAccessExclusive)
+            .context(DeviceInit {})?;
+    }
+
+    Ok(())
+}
+
+fn print_status(_: &ArgMatches) -> Result<()> {
+    let tlkm = TLKM::new().context(TLKMInit {})?;
+    let devices = tlkm.device_enum(&HashMap::new()).context(TLKMInit)?;
+    for x in devices {
+        println!("Device {}", x.id());
+        println!("{:?}", x.status());
+    }
+    Ok(())
+}
+
+fn run_arrayinit(_: &ArgMatches) -> Result<()> {
+    let tlkm = TLKM::new().context(TLKMInit {})?;
+    let devices = tlkm.device_enum(&HashMap::new()).context(TLKMInit)?;
+    for mut x in devices {
+        x.change_access(tapasco::tlkm::tlkm_access::TlkmAccessExclusive)
+            .context(DeviceInit {})?;
+        let mut pe = x.acquire_pe(11).context(DeviceInit)?;
```

Comment:
Using magic numbers can be confusing.
Create a constant integer with a value at the top:
```
       const ARRAY_PE_ID: u32 = 1745;
```
```suggestion
        let mut pe = x.acquire_pe(ARRAY_PE_ID).context(DeviceInit)?;
```

Reply:
Yes, I am not a fan of these numbers as well. For the release I will replace them completely. Then you can request the ID from the API, something like: `device.getPEId("ARRAY_INIT")`. This also makes moving between bitstreams generated by different people (maybe using different IDs) easier.